pkg/hpa/service.go

```diff
@@ -765,3 +768,51 @@ func (c *Service) excludeExternalMetric(ctx context.Context, hpa *v2.HorizontalP
 
 	return newHPA
 }
+
+func (c *Service) CheckHpaMetricStatus(ctx context.Context, currenthpa *v2.HorizontalPodAutoscaler) (bool, error) {
+	//currenthpa = currenthpa.DeepCopy()
+	logger := log.FromContext(ctx)
+	if currenthpa == nil {
+		logger.Info("empty HPA passed into status check, ignore")
+		return true, nil
+	}
+
+	if reflect.DeepEqual(currenthpa.Status, v2.HorizontalPodAutoscalerStatus{}) {
+		return false, fmt.Errorf("HPA empty status, switch to emergency mode")
+	}
+
+	if currenthpa.Status.Conditions == nil {
+		return false, fmt.Errorf("HPA empty conditions, switch to emergency mode")
+	}
+
+	if currenthpa.Status.CurrentMetrics == nil {
+		return false, fmt.Errorf("HPA no metrics, switch to emergency mode")
+	}
```

Comment:
Actually, this might be a general issue that maybe HPA hasn't be reconciled yet when `UpdateHPASpecFromTortoiseAutoscalingPolicy` just created the one.
In such cases, Tortoise controller cannot (or shouldn't) calculate the recommendation.

Can we varify those empty stuff at `GetHPAOnTortoise`, and if empty, we abort the reconciliation? And, at the next reconciliation, HPA status is supposed to be filled by HPA controller, and we can reconcile.

```diff
func (c *Service) GetHPAOnTortoise(ctx context.Context, tortoise *autoscalingv1beta3.Tortoise) (*v2.HorizontalPodAutoscaler, bool, error) {
...
	hpa := &v2.HorizontalPodAutoscaler{}
	if err := c.c.Get(ctx, types.NamespacedName{Namespace: tortoise.Namespace, Name: tortoise.Status.Targets.HorizontalPodAutoscaler}, hpa); err != nil {
		return nil, false, fmt.Errorf("failed to get hpa on tortoise: %w", err)
	}

+	if hpa.Status == nil || hpa.Status.Conditions == nil || .... {
+		// Most likely, HPA is just created and not yet handled by HPA controller.
+		return nil, false, nil
+	}

...
}
```

```diff
	hpa, isReady, err = r.HpaService.GetHPAOnTortoise(ctx, tortoise)
	if err != nil {
		logger.Error(err, "failed to get HPA", "tortoise", req.NamespacedName)
		return ctrl.Result{}, err
	}
+	
+	if !isReady {
+		// HPA is correctly fetched, but looks like not ready yet. We won't be able to calculate things correctly, and hence stop the reconciliation here.
+		return ctrl.Result{}, nil
+	}
```

Reply:
Just a quick question @sanposhiho How should we handle the test case where HPA is created? In the current test case, it is updated to the correct recommendation after one reconcile loop in tortoise conditions. Now it should just be the same tortoise since reconcile was aborted and tests only cover one reconcile loop? Would this be lacking from a testing perspective

Reply:
Here, we change the expectation of the reconciliation loop that creates HPA to stop the reconciliation at that point (if HPA hasn't yet updated immediately by HPA controller) and expect the next reconciliation would handle them appropriately. 

So, we can just change the test case; change the test case's tortoise expectation not to get any update around the recommendation.